pkg/cri/sbserver/sandbox_stats_windows.go

```diff
@@ -19,17 +19,397 @@ package sbserver
+	podSandboxStats := &runtime.PodSandboxStats{}
+	podSandboxStats.Windows = &runtime.WindowsPodSandboxStats{}
+	podSandboxStats.Attributes = &runtime.PodSandboxAttributes{
+		Id:          meta.ID,
+		Metadata:    meta.Config.GetMetadata(),
+		Labels:      meta.Config.GetLabels(),
+		Annotations: meta.Config.GetAnnotations(),
+	}
+
+	metrics, containers, err := c.listWindowsMetricsForSandbox(ctx, sandbox)
+	if err != nil {
+		return nil, fmt.Errorf("failed to obtain container stats during podSandboxStats call: %w", err)
+	}
+
+	statsMap, err := convertMetricsToWindowsStats(metrics)
+	if err != nil {
+		return nil, err
+	}
+
+	podCPU, containerStats, err := c.toPodSandboxStats(sandbox, statsMap, containers, timestamp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert container stats during podSandboxStats call: %w", err)
+	}
+	podSandboxStats.Windows.Cpu = podCPU.Cpu
+	podSandboxStats.Windows.Memory = podCPU.Memory
+	podSandboxStats.Windows.Containers = containerStats
+
+	podSandboxStats.Windows.Network = windowsNetworkUsage(ctx, sandbox, timestamp)
+
+	pidCount, err := c.getSandboxPidCount(ctx, sandbox)
+	if err != nil {
+		return nil, err
+	}
+
+	podSandboxStats.Windows.Process = &runtime.WindowsProcessUsage{
+		Timestamp:    timestamp.UnixNano(),
+		ProcessCount: &runtime.UInt64Value{Value: pidCount},
+	}
+
+	c.saveSandBoxMetrics(podSandboxStats.Attributes.Id, podSandboxStats)
+
+	return podSandboxStats, nil
+}
+
+func convertMetricsToWindowsStats(metrics []*types.Metric) (map[string]*wstats.Statistics, error) {
+	statsMap := make(map[string]*wstats.Statistics)
+	for _, stat := range metrics {
+		var podRuntimeStats *wstats.Statistics
+		podStatsData, err := typeurl.UnmarshalAny(stat.Data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract container metrics: %w", err)
+		}
+
+		// extract the metrics if available
+		// podStatsData can be nil for pods that don't have an actual podsandbox
+		// this is the case for hostprocess pods
+		if podStats, ok := podStatsData.(*wstats.Statistics); ok {
+			podRuntimeStats = podStats
+		}
+
+		statsMap[stat.ID] = podRuntimeStats
+	}
+	return statsMap, nil
+}
+
+func (c *criService) toPodSandboxStats(sandbox sandboxstore.Sandbox, statsMap map[string]*wstats.Statistics, containers []containerstore.Container, timestamp time.Time) (*runtime.WindowsContainerStats, []*runtime.WindowsContainerStats, error) {
+	podMetric, ok := statsMap[sandbox.ID]
+	if !ok {
+		return nil, nil, fmt.Errorf("failed to extract container metrics: no pod stats found")
+	}
+
+	podRuntimeStats, err := c.windowsCPUAndMemoryStats(podMetric)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to extract container metrics: %w", err)
+	}
+
+	windowsContainerStats := []*runtime.WindowsContainerStats{}
+	for _, cntr := range containers {
+		containerMetric := statsMap[cntr.ID]
+
+		if containerMetric == nil {
+			// maybe just a warning?
+			return nil, nil, fmt.Errorf("failed find metric ")
```

Comment:
Regardless of log or error lets include the ID for the container we didn't find metrics for. Is this the hostprocess case where they'd always be nil? If so I wonder if we can thread through enough things to see if that ID was a hostprocess pod and ignore, otherwise error/log.

Reply:
There are two parts to this, the pod runtimestats and the rest of the containers. Here we would have already processed the pod container (lines 113-120) where this check isn't in place (since it is expected for HP containers). 

@kevpar pointed out that we could have already warned in the previous https://github.com/containerd/containerd/pull/7099/files#r1087453810 scan through the metrics so won't really need to handle this.  I can see if I can log the warning only if it is not a hpc